tests/kernel/ee/interrupt/dmac_order.cpp

```diff
@@ -0,0 +1,169 @@
+#include <common-ee.h>
+#include <kernel.h>
+#include <ee_regs.h>
+#include <string.h>
+
+#define MAX_DMAC_RESULTS 9
+#define MAX_DMAC_HANDLERS 10
+
+static u32 g_dmacSema = 0;
+static u32 g_dmacHandlers[MAX_DMAC_HANDLERS];
+static u32 g_dmacHandlerCount = 0;
+static u32 g_orderPos = 0;
+static u32 g_order[MAX_DMAC_RESULTS];
+
+static const u32 __attribute__((aligned(16))) g_dummyTransferData[4] = {0, 0, 0, 0};
+
+typedef s32 (*DmacHandlerFunction)(s32);
+
+void appendResult(u32 result) {
+	if(g_orderPos != MAX_DMAC_RESULTS) {
+		iSignalSema(g_dmacSema);
+		g_order[g_orderPos] = result;
+		g_orderPos++;
+	}
+}
+
+template <s32 number>
+s32 continueDmacHandler(s32 channel) {
+	appendResult(number);
+	return 0;
+}
+
+template <s32 number>
+s32 cancelDmacHandler(s32 channel) {
+	appendResult(number);
+	return -1;
+}
+
+void initializeTest() {
+	g_orderPos = 0;
+	g_dmacHandlerCount = 0;
+	{
+		ee_sema_t dmacSemaInfo;
+		memset(&dmacSemaInfo, 0, sizeof(dmacSemaInfo));
+		dmacSemaInfo.init_count = 0;
+		dmacSemaInfo.max_count  = MAX_DMAC_RESULTS;
+		dmacSemaInfo.option     = 0;
+		g_dmacSema = CreateSema(&dmacSemaInfo);
+	}
+}
+
+void addTestDmacHandler(DmacHandlerFunction handler, s32 next) {
+	int handlerId = AddDmacHandler(DMAC_TO_SPR, handler, next);
+	g_dmacHandlers[g_dmacHandlerCount] = handlerId;
+	g_dmacHandlerCount++;
+}
+
+void removeTestDmacHandler(int index) {
```

Comment:
I agree it's a bit confusing, but I wanted `removeTestDmacHandler` to remove the n-th handler in the array. Doing the removal by id would require an additional step to go through the array and find where it's located.

But if you think it's better, I don't mind changing it so `addTestDmacHandler` returns an id and `removeTestDmacHandler` removes the handler with specified id from the array.

I also wish I could use a map or something here, because implementing the array management functions is no fun. The STL doesn't seem to work very well in PS2SDK, so maybe it would be interesting to build some simple containers at some point?